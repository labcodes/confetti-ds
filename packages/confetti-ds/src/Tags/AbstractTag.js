import React from "react";
import PropTypes from "prop-types";
import { isUndefined } from "lodash";

export default class AbstractTag extends React.Component {
  static propTypes = {
    /** Tag's left element (optional and customizable) */
    renderPrefix: PropTypes.object,
    /** Components to be rendered at the end of the Tag. */
    renderSuffix: PropTypes.object,
    /** Defines a class name to create a custom style for the component. */
    className: PropTypes.string,
    /** This is the Tag's text. */
    text: PropTypes.string.isRequired,
    /** Sets Tag's color. */
    color: PropTypes.string,
    /** Skin of the the rendered Tag. */
    skin: PropTypes.string,
    /** Sets an outline style. */
    isOutline: PropTypes.bool,
    /** Disables the Tag. Will be read by screen readers. When true, will override `disabled`. */
    ariaDisabled: PropTypes.bool,
    /** Disables the Tag. Won't be read by screen readers. */
    disabled: PropTypes.bool,
    /** Action to be executed when the Tag is clicked. */
    onClick: PropTypes.func,

    /** Tab index to manage itens position on screen */
    tabIndex: PropTypes.string,
  };

  static defaultProps = {
    renderPrefix: undefined,
    renderSuffix: undefined,
    className: "",
    isOutline: false,
    color: undefined,
    skin: "pale",
    disabled: false,
    ariaDisabled: false,
    onClick: () => {},
    tabIndex: "0",
  };

  handleEvent = (event) => {
    if ((event.keycode || event.which) === 32) {
      event.preventDefault();
    }
    const { onClick } = this.props;
    if (!isUndefined(onClick)) {
      onClick(event);
    }
  };

  render() {
    const {
      text,
      disabled,
      ariaDisabled,
      isOutline,
      skin,
      color,
      renderPrefix,
      renderSuffix,
      className,
      tabIndex,
    } = this.props;

    return (
      <span
        className={
          `lab-tag ${className}` +
          `${disabled || ariaDisabled ? ` lab-tag--disabled` : ""}` +
          `${isOutline ? ` lab-tag--outline` : ""}` +
          `${color ? ` lab-tag--${color}-${skin}` : ` lab-tag--${skin}`}`
        }
        onClick={!(ariaDisabled || disabled) ? this.handleEvent : () => {}}
        onKeyPress={this.handleEvent}
        role="button"
        tabIndex={tabIndex}
      >
        {renderPrefix}
        {text}
        {renderSuffix}
      </span>
    );
  }
}
