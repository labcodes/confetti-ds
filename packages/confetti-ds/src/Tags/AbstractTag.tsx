import React, { useRef, useEffect, SyntheticEvent, MutableRefObject } from "react";
import { isUndefined } from "lodash";

interface AbstractTagProps {
  /** This is the Tag's text. */
  text: string;
  /** Tag's left element (optional and customizable) */
  renderPrefix?: JSX.Element;
  /** Components to be rendered at the end of the Tag. */
  renderSuffix?: JSX.Element;
  /** Defines a class name to create a custom style for the component. */
  className?: string;
  /** Sets Tag's color. */
  color?: string;
  /** Skin of the rendered Tag. */
  skin?: string;
  /** Sets an outline style. */
  isOutline?: boolean;
  /** Disables the Tag. Will be read by screen readers. When true, will override `disabled`. */
  ariaDisabled?: boolean;
  /** Disables the Tag. Won't be read by screen readers. */
  disabled?: boolean;
  /** Action to be executed when the Tag is clicked. */
  onClick?: (event?: SyntheticEvent) => any;
  /** tabIndex is used to define the navigation order for focusable elements. If not undefined, it is passed to the tag component. */
  tabIndex?: number;
  /** This prop is a boolean to verify if the tag is a child of the TagDropdown component */
  isDropdown?: boolean;
  /** This function is used to handle click or keydown interactions */
  onInteraction?: ({
    event,
    ref,
  }: {
    event: any;
    ref: React.MutableRefObject<boolean>;
  }) => any;
  /** This is the TagDropdownItem value. It is used TagDropdown options */
  value?: string;
  /**  This prop is used to verify if the option is selected */
  isSelected?: boolean;
  /**   This is the TagDropdownItem or TagDropdownTrigger role. It is used to let disabled people know the type of the element. */
  role?: string;
  /** This function is used on AbstractTag to set the current Ref */
  setRef?: (ref: MutableRefObject<HTMLButtonElement>) => any;
}

export default function AbstractTag({
  text,
  color,
  renderPrefix,
  renderSuffix,
  className,
  tabIndex,
  isOutline = false,
  skin = "pale",
  disabled = false,
  ariaDisabled = false,
  onClick = () => {},
  onInteraction = () => {},
  setRef = () => {},
  isDropdown = false,
  isSelected = false,
  value = "",
  role = "",
}: AbstractTagProps) {
  const ref = useRef();

  const handleEvent = (event) => {
    if ((event.key || event.which) === 32) {
      event.preventDefault();
    }
    if (!isUndefined(onClick)) {
      onClick(event);
    }

    if (!isUndefined(onInteraction)) {
      onInteraction({ event, ref });
    }
  };

  useEffect(() => {
    if (isDropdown) {
      setRef(ref);
    }
  }, [isDropdown]);

  if (isDropdown)
    return (
      <button
        type="button"
        role={role}
        aria-selected={isSelected}
        className={
          `lab-tag ${className}` +
          `${disabled || ariaDisabled ? ` lab-tag--disabled` : ""}` +
          `${isOutline ? ` lab-tag--outline` : ""}` +
          `${color ? ` lab-tag--${color}-${skin}` : ` lab-tag--${skin}`}`
        }
        onClick={!(ariaDisabled || disabled) ? handleEvent : () => {}}
        onKeyDown={handleEvent}
        tabIndex={tabIndex}
        ref={ref}
        disabled={disabled || ariaDisabled}
        id={value}
        value={value}
      >
        {renderPrefix}
        {text}
        {renderSuffix}
      </button>
    );

  return (
    <span
      className={
        `lab-tag ${className}` +
        `${disabled || ariaDisabled ? ` lab-tag--disabled` : ""}` +
        `${isOutline ? ` lab-tag--outline` : ""}` +
        `${color ? ` lab-tag--${color}-${skin}` : ` lab-tag--${skin}`}`
      }
      onClick={!(ariaDisabled || disabled) ? handleEvent : () => {}}
      onKeyDown={handleEvent}
      role="button"
      tabIndex={tabIndex}
    >
      {renderPrefix}
      {text}
      {renderSuffix}
    </span>
  );
}
