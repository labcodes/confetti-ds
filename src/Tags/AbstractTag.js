import React from "react";
import PropTypes from "prop-types";
import { isUndefined } from "lodash";

export default class AbstractTag extends React.Component {
  static propTypes = {
    /** Tag's left element (optional and customizable) */
    renderPrefix: PropTypes.object,
    /** Components to be rendered at the end of the Tag. */
    renderSuffix: PropTypes.object,
    /** Defines a class name to create a custom style for the component. */
    className: PropTypes.string,
    /** This is the Tag's text. */
    text: PropTypes.string.isRequired,
    /** Sets Tag's color. */
    color: PropTypes.string,
    /** Skin of the the rendered Tag. */
    skin: PropTypes.string,
    /** Sets an outline style. */
    isOutline: PropTypes.bool,
    /** Disables the Tag. Will be read by screen readers. When true, will override `disabled`. */
    ariaDisabled: PropTypes.bool,
    /** Disables the Tag. Won't be read by screen readers. */
    disabled: PropTypes.bool,
    /** Action to be executed when the Tag is clicked. */
    onClick: PropTypes.func,
    /** tabIndex is used to define the navigation order for focusable elements. If not undefined, it is passed to the tag component. */
    tabIndex: PropTypes.string,
    /** This prop is a boolean to verify if the tag is a child of the TagDropdown component */
    isDropdown: PropTypes.bool,
    /** This prop sets the TagDropdownItem ref for option and triggers in a TagDropdown component. */
    setRef: PropTypes.func,
    /** This function is used to handle click or keydown interactions */
    onInteraction: PropTypes.func,
    /** This is the TagDropdownItem value. It is used TagDropdown options */
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    /**  This prop is used to verify if the option is selected */
    isSelected: PropTypes.bool,
    /**   This is the TagDropdownItem or TagDropdownTrigger role. It is used to let disabled people know the type of the element. */
    role: PropTypes.string,
  };

  static defaultProps = {
    renderPrefix: undefined,
    renderSuffix: undefined,
    className: "",
    isOutline: false,
    color: undefined,
    skin: "pale",
    disabled: false,
    ariaDisabled: false,
    onClick: () => {},
    onInteraction: () => {},
    setRef: () => {},
    tabIndex: undefined,
    isDropdown: false,
    isSelected: false,
    value: "",
    role: "",
  };

  constructor(props) {
    super(props);

    if (props.isDropdown) this.ref = React.createRef();
  }

  componentDidMount() {
    const { setRef, isDropdown } = this.props;
    if (isDropdown) setRef(this.ref);
  }

  handleEvent = (event) => {
    const { onClick, onInteraction } = this.props;

    if ((event.keycode || event.which) === 32) {
      event.preventDefault();
    }
    if (!isUndefined(onClick)) {
      onClick(event);
    }

    if (!isUndefined(onInteraction)) {
      onInteraction({ event, ref: this.ref });
    }
  };

  render() {
    const {
      text,
      disabled,
      ariaDisabled,
      isOutline,
      skin,
      color,
      renderPrefix,
      renderSuffix,
      className,
      tabIndex,
      isDropdown,
      value,
      isSelected,
      role,
    } = this.props;

    if (isDropdown)
      return (
        <button
          type="button"
          role={role}
          aria-selected={isSelected}
          className={
            `lab-tag ${className}` +
            `${disabled || ariaDisabled ? ` lab-tag--disabled` : ""}` +
            `${isOutline ? ` lab-tag--outline` : ""}` +
            `${color ? ` lab-tag--${color}-${skin}` : ` lab-tag--${skin}`}`
          }
          onClick={!(ariaDisabled || disabled) ? this.handleEvent : () => {}}
          onKeyDown={this.handleEvent}
          tabIndex={tabIndex}
          ref={this.ref}
          disabled={disabled || ariaDisabled}
          id={value}
          value={value}
        >
          {renderPrefix}
          {text}
          {renderSuffix}
        </button>
      );

    return (
      <span
        className={
          `lab-tag ${className}` +
          `${disabled || ariaDisabled ? ` lab-tag--disabled` : ""}` +
          `${isOutline ? ` lab-tag--outline` : ""}` +
          `${color ? ` lab-tag--${color}-${skin}` : ` lab-tag--${skin}`}`
        }
        onClick={!(ariaDisabled || disabled) ? this.handleEvent : () => {}}
        onKeyPress={this.handleEvent}
        role="button"
        tabIndex={tabIndex}
      >
        {renderPrefix}
        {text}
        {renderSuffix}
      </span>
    );
  }
}
