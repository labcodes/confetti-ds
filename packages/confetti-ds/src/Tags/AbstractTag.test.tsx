import { shallow } from "enzyme";
import React from "react";
import renderer from "react-test-renderer";

import AbstractTag from "./AbstractTag";

describe("AbstractTag", () => {
  it("renders with base props for 'simple' type", async () => {
    const renderedComponent = renderer
      .create(<AbstractTag text="Test SimpleTag" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders with base props for 'removable' type", async () => {
    const renderedComponent = renderer
      .create(
        <AbstractTag className="lab-tag--removable" text="Test RemovableTag" />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders with base props for 'togglable' type", async () => {
    const renderedComponent = renderer
      .create(
        <AbstractTag className="lab-tag--togglable" text="Test TogglableTag" />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders with base props for 'dropdown' type", async () => {
    const renderedComponent = renderer
      .create(
        <AbstractTag className="lab-tag--dropdown" text="Test DropdownTag" />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("triggers onClick", async () => {
    const mockOnClick = jest.fn();
    const shallowComponent = shallow(
      <AbstractTag text="Test SimpleTag" onClick={mockOnClick} />
    );

    expect(mockOnClick).not.toBeCalled();
    shallowComponent.find("span").at(0).simulate("click", { test: "event" });
    expect(mockOnClick).toBeCalledWith({ test: "event" });
  });

  it("doesn't trigger onClick if ariaDisabled", async () => {
    const mockOnClick = jest.fn();
    const shallowComponent = shallow(
      <AbstractTag ariaDisabled text="Test SimpleTag" onClick={mockOnClick} />
    );

    expect(mockOnClick).not.toBeCalled();
    shallowComponent.find("span").at(0).simulate("click");
    expect(mockOnClick).not.toBeCalled();
  });
});
