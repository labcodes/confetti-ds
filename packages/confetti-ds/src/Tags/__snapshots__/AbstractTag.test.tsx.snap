// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AbstractTag renders with base props for 'dropdown' type 1`] = `
<span
  className="lab-tag lab-tag--dropdown lab-tag--pale"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
>
  Test DropdownTag
</span>
`;

exports[`AbstractTag renders with base props for 'removable' type 1`] = `
<span
  className="lab-tag lab-tag--removable lab-tag--pale"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
>
  Test RemovableTag
</span>
`;

exports[`AbstractTag renders with base props for 'simple' type 1`] = `
<span
  className="lab-tag undefined lab-tag--pale"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
>
  Test SimpleTag
</span>
`;

exports[`AbstractTag renders with base props for 'togglable' type 1`] = `
<span
  className="lab-tag lab-tag--togglable lab-tag--pale"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
>
  Test TogglableTag
</span>
`;
